### PR TITLE
feat(cloudfront): CloudFrontOriginAccessIdentity

### PR DESCRIPTION
**Pull request by workeitel**
Example usage:

```
const bucket = new s3.Bucket(stack, 'Bucket');
const oai = new cloudfront.CloudFrontOriginAccessIdentity(stack, 'OAI');

bucket.grantRead(oai); // must be granted explicitly

const dist = new cloudfront.CloudFrontWebDistribution(stack, 'Distribution', {
  originConfigs: [{
    behaviors: [{ isDefaultBehavior: true }],
    s3OriginSource: {
      s3BucketSource: bucket,
      originAccessIdentity: oai,
    },
  }]
});
```

It supports the old `originAccessIdentityId` as well:

```
const dist = new cloudfront.CloudFrontWebDistribution(stack, 'Distribution', {
  originConfigs: [{
    behaviors: [{ isDefaultBehavior: true }],
    s3OriginSource: {
      s3BucketSource: bucket,
      originAccessIdentityId: "...",  // or oai.name
    },
  }]
});
```

And importing Origin Access Identities:

```
const oai = cloudfront.CloudFrontOriginAccessIdentity.fromName(
  stack,
  'OAIImported',
  'AEFASDF234'
);
```

NOTE: while typically the S3CanonicalUserId is used in the policy this
value is not available after import so instead the CloudFront Arn is
constructed manually.

fixes: https://github.com/aws/aws-cdk/issues/941

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cloudfront/test/integ.cloudfront-s3.ts [hunk 1]
```diff
@@ -12,19 +11,23 @@ const oai = new cloudfront.CfnCloudFrontOriginAccessIdentity(stack, 'OAI', {
...
     },
   }]
 });
-bucket.addToResourcePolicy(new iam.PolicyStatement({
-  actions: ['s3:Get*', 's3:List*'],
-  resources: [bucket.bucketArn, bucket.arnForObjects('*')],
-  principals: [new iam.CanonicalUserPrincipal(oai.attrS3CanonicalUserId)],
-}));
+
+bucket.grantRead(oaiImported);
```
Comment (workeitel) [inline packages/@aws-cdk/aws-cloudfront/test/integ.cloudfront-s3.ts]: should this be done implicitly when adding to CloudFront?

File: packages/@aws-cdk/aws-cloudfront/package.json [hunk 2]
```diff
@@ -50,6 +50,11 @@
   "cdk-build": {
     "cloudformation": "AWS::CloudFront"
   },
+  "awslint": {
+    "exclude": [
+      "props-physical-name:@aws-cdk/aws-cloudfront.CloudFrontOriginAccessIdentityProps"
```
Comment (workeitel) [inline packages/@aws-cdk/aws-cloudfront/package.json]: I used just `name` instead of `cloudFrontOriginAccessIdentityName`

For example:
```
const oai = cloudfront.CloudFrontOriginAccessIdentity.fromName(
  stack,
  'OAIImported',
  'AEFASDF234'
);
```

instead of

```
const oai = cloudfront.CloudFrontOriginAccessIdentity.fromCloudFrontOriginAccessIdentityName(
  stack,
  'OAIImported',
  'AEFASDF234'
);
```

File: packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts [hunk 3]
```diff
@@ -244,6 +245,11 @@ export interface S3OriginConfig {
    * The optional ID of the origin identity cloudfront will use when calling your s3 bucket.
    */
   readonly originAccessIdentityId?: string;
```
Comment (workeitel) [inline packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts]: Should this be dropped and instead be a breaking change?

Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts]: yes please. It's an experimental module, so you can break.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+export interface ICloudFrontOriginAccessIdentity extends cdk.IResource, iam.IGrantable  {
+  /**
+   * The Origin Access Identity Name
+   */
+  readonly name: string;
+}
+
+export abstract class BaseCloudFrontOriginAccessIdentity extends cdk.Resource {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: No need to export the base class

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+   * The Arn to include in S3 bucket policy to allow CloudFront access
+   */
+  protected arn(): string {
+    return cdk.Stack.of(this).formatArn(
+      {
+        service: "iam",
+        region: "", // global
+        account: "cloudfront",
+        resource: "user",
+        resourceName: `CloudFront Origin Access Identity ${this.name}`
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: what? is that the resource name?

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      }
+    );
+  }
+}
+
+export interface CloudFrontOriginAccessIdentityAttributes {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Why is this interface needed?

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  readonly name: string;
+}
+
+/**
+ * An origin access identity is a special CloudFront user that you can
+ * associate with Amazon S3 origins, so that you can secure all or just some of
+ * your Amazon S3 content.
+ */
+export class CloudFrontOriginAccessIdentity extends BaseCloudFrontOriginAccessIdentity implements ICloudFrontOriginAccessIdentity {
+  public static fromCloudFrontOriginAccessIdentityAttributes(
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: This method is not needed

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    name: string): BaseCloudFrontOriginAccessIdentity {
+    return CloudFrontOriginAccessIdentity.fromCloudFrontOriginAccessIdentityAttributes(scope, id, {
+      name
+    });
+  }
+
+  /**
+   * CDK L1 resource
+   */
+  public readonly resource: CfnCloudFrontOriginAccessIdentity;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: We do not expose L1s as public API.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+  /**
+   * The Amazon S3 canonical user ID for the origin access identity, used when
+   * giving the origin access identity read permission to an object in Amazon
+   * S3.
+   *
+   * @attribute S3CanonicalUserId
+   */
+  public readonly s3CanonicalUserId: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Attributes must begin with the resource name

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+  /**
+   * Derived principal value for bucket access
+   */
+  public readonly grantPrincipal: iam.IPrincipal;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: why is this needed as a public API?

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 4 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Build fails otherwise with if I don't export it

```
JSII: Cannot use private type BaseCloudFrontOriginAccessIdentity in exported declaration
```

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 5 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: yes, it's one of the two ways:

Either by S3 CanonicalUser or via this ARN: https://stackoverflow.com/questions/44877359/aws-where-do-i-see-a-list-of-origin-access-identities

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 6 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: I thought `fromCloudFrontOriginAccessIdentityAttributes` was a convention - removed both

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 7 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: I thought this is a convention - removed it

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 8 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: made it private

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 9 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: I was hoping it's not required since the name is super long `this.cloudFrontOriginAccessIdentityS3CanonicalUserId` - but sure, changed it.

Does that apply to `Name` as well? https://github.com/aws/aws-cdk/pull/4491/files#r334299183

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 10 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: When I understand it correctly it's part of `iam.IGrantable` and required for `bucket.grantRead(oai)`

File: packages/@aws-cdk/aws-cloudfront/test/integ.cloudfront-s3.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/test/integ.cloudfront-s3.ts]: yes

File: packages/@aws-cdk/aws-cloudfront/package.json (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/package.json]: Our convention is to use the full name (See other L2s)

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+   * @default "Allows CloudFront to reach the bucket"
+   */
+  readonly comment?: string;
+}
+
+export interface ICloudFrontOriginAccessIdentity extends cdk.IResource, iam.IGrantable  {
+  /**
+   * The Origin Access Identity Name
+   */
+  readonly name: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: This this is an attribute, we prefix it with the name of the resource by convention: `originAccessIdentityName` 
Also, it needs an @attribute tag..

Can you make sure you remove all the awslint exclusions for this resource from package.json? You should get quite a lot of guidance during build as a result.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
+import iam = require('@aws-cdk/aws-iam');
+import cdk = require('@aws-cdk/core');
+import { CfnCloudFrontOriginAccessIdentity } from './cloudfront.generated';
+
+export interface CloudFrontOriginAccessIdentityProps {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Should be called OriginAccessIdentity (no CloudFront prefix). I know this module has a lot of these, but those are non-conventional and will be changed. Let's build this one per our guidelines.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 4 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: This is because you use this as a return type in the `fromName` method, which should return the interface. Remove export and rename to `OriginAccessIdentityBase`

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    );
+  }
+}
+
+/**
+ * An origin access identity is a special CloudFront user that you can
+ * associate with Amazon S3 origins, so that you can secure all or just some of
+ * your Amazon S3 content.
+ */
+export class CloudFrontOriginAccessIdentity extends BaseCloudFrontOriginAccessIdentity implements ICloudFrontOriginAccessIdentity {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Rename to `OriginAccessIdentity`

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+  public static fromName(
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: `fromOriginAccessIdentityName`

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    scope: cdk.Construct,
+    id: string,
+    name: string): BaseCloudFrontOriginAccessIdentity {
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Oh, that's why jsii errors when you try to make this private. This should return an `IOriginAccessIdentity`, not a concrete class.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+   * CDK L1 resource
+   */
+  private readonly resource: CfnCloudFrontOriginAccessIdentity;
+
+  constructor(scope: cdk.Construct, id: string, props?: CloudFrontOriginAccessIdentityProps) {
+    super(scope, id);
+
+    this.resource = new CfnCloudFrontOriginAccessIdentity(this, "Resource", {
+      cloudFrontOriginAccessIdentityConfig: {
+        comment: (props && props.comment) || "Allows CloudFront to reach to the bucket"
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: "to reach the bucket", no? Is it required to add a comment? If no, let's just leave it empty by default.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+      }
+    });
+    // physical id - OAI name
+    this.name = this.resource.ref.toString();
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Please see other examples on how physical names should be handled. It's a bit convoluted but required in order to support things like cross-account access.

File: packages/@aws-cdk/aws-cloudfront/package.json (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/package.json]: Please remove all excludes related to this resource.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 16 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Fixed the spelling. The comment is required: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-cloudfrontoriginaccessidentity-cloudfrontoriginaccessidentityconfig.html

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 17 above)
Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: I did not find another example of a physical id which can't be specified in properties on create and is handled differently. All the vpc-id, subnet-id resources use the same method of not specify a `name` attribute in properties and excluding `"props-physical-name:` in awslint. It would be great if you could point me to a better example.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts (hunk 10 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Yes you are right. sorry.

File: packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+  constructor(scope: cdk.Construct, id: string, props?: OriginAccessIdentityProps) {
+    super(scope, id);
+
+    this.resource = new CfnCloudFrontOriginAccessIdentity(this, "Resource", {
+      cloudFrontOriginAccessIdentityConfig: {
+        comment: (props && props.comment) || "Allows CloudFront to reach the bucket"
+      }
+    });
+    // physical id - OAI name
+    this.originAccessIdentityName = this.resource.ref;
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: please apply our pattern for physical IDs. 

- call super constructor with `physicalName` ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L993))
- use `this.physicalName` when passed to L1 ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1001)
- use `this.getResourceNameAttribute` and `this.getResourceArnAttribute` to render the name and ARN attributes ([example](https://github.com/aws/aws-cdk/blob/652a8f5b4d253babb57dfb3f900794071509a565/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1017))./

I am aware this is not documented anywhere and we haven't found a way to enforce it.

Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/lib/origin_access_identity.ts]: Thanks for the help. The `physicalName` is the origin access id which is returned when creating OAI resource. It is not known before.

> 1) call super constructor with `physicalName`

The constructor does not get any physicalName so I can't pass it to `super`. But I added it at least to the `Import` class.

> 2) use `this.physicalName` when passed to L1

The `physicalName` is never passed to the L1 because it's not known before.

> 3) use `this.getResourceNameAttribute` and `this.getResourceArnAttribute` to render the name and ARN attributes

I used the `getResourceNameAttribute` but since OAIs don't have ARNs I can't use `getResourceArnAttribute`.

File: packages/@aws-cdk/aws-cloudfront/package.json [hunk 19]
```diff
@@ -92,6 +92,7 @@
   "stability": "experimental",
   "awslint": {
     "exclude": [
+      "props-physical-name:@aws-cdk/aws-cloudfront.OriginAccessIdentityProps",
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudfront/package.json]: why?

Comment (workeitel) [reply packages/@aws-cdk/aws-cloudfront/package.json]: Because there is no physical name for the constructor but awslint always was a physical name in the props

File: packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts (hunk 3 above)
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts]: @eladb it was not included as a breaking change in the release notes and just broke every website I am running ...

Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts]: Sorry! Apologies for the bad experience. I'll update the release notes in the changelog and on GitHub.

Comment (eladb) [reply packages/@aws-cdk/aws-cloudfront/lib/web_distribution.ts]: Updated. Thanks for letting us know @hoegertn and sorry again for the hassle.
